services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,202 @@
+        CompletableFuture<InfoData> future = new CompletableFuture<>();
+        pendingRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<List<InfoData>> sendAndGetChunked(String topic, Message message) {
+        CompletableFuture<List<InfoData>> future = new CompletableFuture<>();
+        pendingChunkedRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Received invalid message: {}", message);
+            return;
+        }
+
+        if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+            throw new MessageException(error);
+        }
+
+        if (message instanceof ChunkedInfoMessage) {
+            processChunkedMessage((ChunkedInfoMessage) message);
+        } else if (pendingRequests.containsKey(message.getCorrelationId())) {
+            InfoMessage infoMessage = (InfoMessage) message;
+            CompletableFuture<InfoData> future = pendingRequests.remove(message.getCorrelationId());
+            future.complete(infoMessage.getData());
+        }
+    }
+
+    /**
+     * Performs searching and collecting all chunked messages into one chain if possible.
+     */
+    private synchronized void processChunkedMessage(ChunkedInfoMessage received) {
+        List<ChunkedInfoMessage> chain;
+        if (messagesChains.containsKey(received.getCorrelationId())) {
+            chain = messagesChains.remove(received.getCorrelationId());
+            chain.add(received);
+        } else if (pendingChunkedRequests.containsKey(received.getCorrelationId())) {
+            chain = Lists.newArrayList(received);
+        } else {
+            unlinkedChunks.put(received.getCorrelationId(), received);
+            return;
+        }
+
+        ChunkedInfoMessage tail = received;
+        // Check whether next messages for current chain is already received. If yes - add them into that chain.
+        while (tail.getNextRequestId() != null && unlinkedChunks.containsKey(tail.getNextRequestId())) {
```

Comment:
Recursion can be here, you can call yourself to handle unlinked chains. :)

Reply:
In case of huge amount of chunked messages there will be an error